crates/shared/src/event_handling.rs

```diff
@@ -161,24 +263,162 @@ where
         if !have_deleted_old_events {
             self.store.replace_events(Vec::new(), range.clone()).await?;
         }
-        self.last_handled_block = Some(range.end().to_u64());
+
+        let blocks = self
+            .block_retriever
+            .blocks(RangeInclusive::new(
+                range.end().saturating_sub(MAX_REORG_BLOCK_COUNT),
+                *range.end(),
+            ))
+            .await?;
+
+        self.update_last_handled_blocks(&blocks);
+        Ok(())
+    }
+
+    async fn update_events_from_latest_blocks(
+        &mut self,
+        blocks: &[BlockNumberHash],
+        is_reorg: bool,
+    ) -> Result<()> {
+        let (blocks, events) = self.past_events_by_block_hashes(blocks).await;
+        track_block_range(&format!("range_{}", blocks.len()));
+        tracing::debug!(
+            "final blocks for updating: {:?} - {:?}",
+            blocks.first(),
+            blocks.last()
+        );
+        if blocks.is_empty() {
+            return Ok(());
+        }
+        let range = RangeInclusive::new(blocks.first().unwrap().0, blocks.last().unwrap().0);
+        if is_reorg {
+            self.store.replace_events(events, range.clone()).await?;
+        } else {
+            self.store.append_events(events).await?;
+        }
+
+        self.update_last_handled_blocks(&blocks);
         Ok(())
     }
 
-    async fn past_events(
+    async fn past_events_by_block_hashes(
         &self,
-        block_range: &RangeInclusive<BlockNumber>,
+        blocks: &[BlockNumberHash],
+    ) -> (Vec<BlockNumberHash>, Vec<EthcontractEvent<C::Event>>) {
+        let (mut blocks_filtered, mut events) = (vec![], vec![]);
+        for (i, result) in future::join_all(
+            blocks
+                .iter()
+                .map(|block| self.contract.get_events().block_hash(block.1).query()),
+        )
+        .await
+        .into_iter()
+        .enumerate()
+        {
+            match result {
+                Ok(e) => {
+                    blocks_filtered.push(blocks[i]);
+                    events.extend(e.into_iter());
+                }
+                Err(_) => return (blocks_filtered, events),
```

Comment:
Is it really correct to return early if one of the calls errors? Shouldn't the whole function error? It seems like on error we would insert partial events without noticing. I think maybe you're trying to account for this in `update_events_from_latest_blocks` but it is not obvious to me how that works. Need at least comment explaining correct.

Even on no error I don't think it is correct because the blocks we are are returning here in blocks_filtered aren't sorted. We might put them into the vector in any order.

Reply:
Isn't the order guaranteed by the `future::join_all`? At least docs say that: https://docs.rs/futures/0.2.1/futures/future/fn.join_all.html
Or I might be poorly understanding this function (unfortunately not so easy to test this out with UTs right now)

> Is it really correct to return early if one of the calls errors?

Yes, the point is to be able to return partial results (but ordered) because it may happen to fail on the last call for example, because of the node propagation delay.
But in most cases, I expect to have only one block here.

Will add comment.

Reply:
You're right that the order is guaranteed.